tests/bus_test.go

```diff
@@ -487,6 +487,161 @@ func TestSendingPanic(t *testing.T) {
 	}
 }
 
+func TestEmptyBody(t *testing.T) {
+	b := createNamedBusForTest(testSvc5)
+	proceed := make(chan bool)
+	b.SetGlobalRawMessageHandler(func(tx *sql.Tx, delivery *amqp.Delivery) error {
+		proceed <- true
+		return nil
+	})
+
+	err := b.Start()
+	if err != nil {
+		t.Errorf("could not start bus for test error: %s", err.Error())
+	}
+	defer func() {
+		err := b.Shutdown()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	conn, err := amqp.Dial(connStr)
+	if err != nil {
+		t.Error("couldnt connect to rabbitmq")
+	}
+	defer conn.Close()
+
+	ch, err := conn.Channel()
+	if err != nil {
+		t.Error("couldnt open rabbitmq channel for publishing")
+	}
+	defer ch.Close()
+
+	cmd := amqp.Publishing{}
+	err = ch.Publish("", testSvc5, true, false, cmd)
+	if err != nil {
+		t.Error("couldnt send message on rabbitmq channel")
+	}
+
+	select {
+	case <-proceed:
+		fmt.Println("success")
+	case <-time.After(2 * time.Second):
+		t.Fatal("timeout, failed to consume message with missing body")
+	}
+}
+
+func TestDeadEmptyBody(t *testing.T) {
+	b := createBusWithConfig(testSvc5, "grabbit-dead", true, true,
+		gbus.BusConfiguration{MaxRetryCount: 0, BaseRetryDuration: 0})
+
+	proceed := make(chan bool)
+	b.HandleDeadletter(func(tx *sql.Tx, delivery *amqp.Delivery) error {
+		proceed <- true
+		return nil
+	})
+
+	err := b.Start()
+	if err != nil {
+		t.Errorf("could not start bus for test error: %s", err.Error())
+	}
+
+	defer func() {
+		err := b.Shutdown()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	conn, err := amqp.Dial(connStr)
+	if err != nil {
+		t.Error("couldnt connect to rabbitmq")
+	}
+	defer conn.Close()
+	ch, err := conn.Channel()
+	if err != nil {
+		t.Error("couldnt open rabbitmq channel for publishing")
+	}
+	defer ch.Close()
+
+	headersMap := make(map[string]interface{})
+	headersMap["x-death"] = make([]interface{}, 0)
+	cmd := amqp.Publishing{Headers: headersMap}
+	err = ch.Publish("", testSvc5, true, false, cmd)
+	if err != nil {
+		t.Error("couldnt send message on rabbitmq channel")
+	}
+
+	select {
+	case <-proceed:
+		fmt.Println("success")
+	case <-time.After(2 * time.Second):
+		t.Fatal("timeout, failed to consume message with missing body")
+	}
+}
+
+func TestFailHandlerInvokeOfMessageWithEmptyBody(t *testing.T) {
+	rejected, _ := metrics.GetRejectedMessagesValue()
+	b := createBusWithConfig(testSvc5, "grabbit-dead1", true, true,
+		gbus.BusConfiguration{MaxRetryCount: 0, BaseRetryDuration: 0})
+
+	proceed := make(chan bool)
+	b.HandleDeadletter(func(tx *sql.Tx, delivery *amqp.Delivery) error {
+		proceed <- true
+		return nil
+	})
+	err := b.HandleMessage(&Command3{}, func(invocation gbus.Invocation, message *gbus.BusMessage) error {
```

Comment:
try reverting back to use Command2 and see if it still raises that strange error. if there is some sort of fundamental bug I want us to identify it and fix. I am hoping that my prev PR fixes this issue as well.

Reply:
when reverting back to svc1 it works.
 But when reverting back to "Command1" I get :

```--- FAIL: TestFailHandlerInvokeOfMessageWithEmptyBody (0.01s)
panic: gob: registering duplicate names for *tests.Command1: "github.com/wework/grabbit/tests.Command1" != "*tests.Command1" [recovered]
	panic: gob: registering duplicate names for *tests.Command1: "github.com/wework/grabbit/tests.Command1" != "*tests.Command1"
```

Reply:
@adiweiss  A few obsurvations:

When running only the TestFailHandlerInvokeOfMessageWithEmptyBody test with **&Command1** it passes
go test ./tests -run=TestFailHandlerInvokeOfMessageWithEmptyBody 
But when running the entire test suite it fails with the above gob error.

When changing the test code to
b.HandleMessage(**Command1**,....) passing in the struct instead of pointer the test suite runs successfully.

Reply:
well, I changed it and now the test pass. but still remains the question - why does it work when running just this test?!

Reply:
seems like what is going on is that the gob package has a bug that is maintained for backward compatibility.
see: [encoding/gob/type.go#L847](https://github.com/golang/go/blob/87113f7eadf6d8b12279709f05c0359b54b194ea/src/encoding/gob/type.go#L847)
and
[encoding/gob/type.go#L824](https://github.com/golang/go/blob/87113f7eadf6d8b12279709f05c0359b54b194ea/src/encoding/gob/type.go#L824)

So what is happening is that since all the tests run in the same context and the registered types in gob get carried over between test runs + the bug in gob then &Command1{} and Command1{} gets transformed to the same user type causes the panic.

you can reproduce using the following test

```
func TestGob(t *testing.T) {

	m1 := &Command1{}
	m2 := Command1{}

	gob.Register(m1)
	gob.Register(m2)
}
```